streams/shared/src/main/scala/zio/stream/experimental/ZStream.scala

```diff
@@ -115,6 +182,36 @@ class ZStream[-R, +E, -M, +B, +A](
       _ <- pull.toManaged_
     } yield ()
 
+  /**
+   * Runs the sink on the stream to produce either the sink's internal state or an error.
+   */
+  def runQuery[R1 <: R, E1 >: E, A1 >: A, B](sink: ZSink[R1, E1, B, A1, Nothing]): ZIO[R1, E1, B] =
+    (self.process <*> sink.process).use {
+      case (command, control) =>
+        def pull: ZIO[R1, E1, B] =
+          command.pull.foldM(
```

Comment:
Love this implementation. The new encoding looks to be paying off so far!

Reply:
Yes, so clear!